### PR TITLE
Progress on e2e lowering

### PR DESCRIPTION
**Pull request by silvasean**
Recommended review order: commit by commit.

Two main changes:
- pass that adjusts calling conventions to incorporate torch.type_bound (and eliminate dummy NoneType returns + tuples eventually)
- primitive ArrayToTensor pass

Next steps are to write shape/dtype refinement stuff and then we can run on refbackend (immediately after will pull in IREE an run e2e with that).

### CODE REVIEW

File: lib/Dialect/Numpy/IR/NumpyOps.cpp [hunk 1]
```diff
@@ -77,6 +77,49 @@ bool StaticInfoCastOp::areCastCompatible(mlir::TypeRange inputs,
...
+  // -->
+  // create_array_from_tensor(tensor_static_info_cast(%tensor))
+  //
+  // This pattern tends to create more tensor code and less array code.
+  // This form is considered more canonical because it has same number of ops
+  // but is more analyzable.
+  //
+  // TODO: Consider a world where we numpy.ndarray can track an "immutable" bit
+  // which makes it tensor-like. Is that useful?
+  patterns.add(+[](StaticInfoCastOp op, PatternRewriter &rewriter) {
```
Comment (stellaraccident) [inline lib/Dialect/Numpy/IR/NumpyOps.cpp:92]: Do you need the unary '+' in this case? (I think I recall that it becomes optional in newer standards)

**Review by stellaraccident — APPROVED** [review]
This is quite nice and great to see this step being taken.

The conversion code is indeed kind of gross. It LGTM modulo it being great to get River's eyes on this for recommendations/enhancements in the future.

File: lib/Dialect/Numpy/IR/NumpyOps.cpp (hunk 1 above)
Comment (silvasean) [reply lib/Dialect/Numpy/IR/NumpyOps.cpp:92]: Needed on my machine sadly.

**Comment by silvasean** [comment]
Will talk to River about it. Thanks!

**Pull request merged**